File: js/e4.js
/* Akachukwu Obi, 2018

Project Euler #4 - Largest palindrome product

A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99. Find the largest palindrome made from the product of two 3-digit numbers. (do these numbers have to be different?)

Retrieved June 25, 2018 from https://projecteuler.net/problem=4
*/

/* ---------------------------------------------//
First attempts at solving this problem

Theorem: all four sized palindromic numbers (numbers between 1000 and 10000) are divisible by 11. Check out this program: 

// first, we define divisibility by 11 as having no remainders after 11 divides the number; i.e. number % 11 === 0

    var str, set, i, j;
    set = [];
    for (i = 1000; i < 10000; i += 1) {
        str = String(i);
        if (str.charAt(0) === str.charAt(str.length - 1) && str.charAt(1) === str.charAt(str.length - 2)) {
            set.push(i);
        }
    }
    for (j = 0; j < set.length; j += 1) {
        if (set[j] % 11 !== 0) {
            console.log(set[j], "is not a factor of 11");
        }
    }
    console.log(set);
    
Conjecture: all even-sized palindromic numbers are divisible by 11.

This conjecture might be helpful because we can find a small 6-digit palindrome that is a product of two three-digit numbers: 111111 = 143 * 777. This means that the largest palindrome number with two three digit multiples is 6-digits long. 

Revised steps and pseudocode:
1. While (998001 > palindrome > 100001), 
    get3DigitFactors (return factors) 
        for (count < 3digFactors.length / 2) only check half
            if 3digfactors[count] * 3digfactors[count++] === palindrome
                break
            else 
                go checkPalindrome(palindrome[index++])
            return palindrome

To cut my calculation time, I could start with the largest number and progress from there.

// -------------------------------------------- */

function getThreeDigitsFactors(number) {
    'use strict';
    var factors, threeDigitFactors, sqrt, possibleFactor, otherPossibleFactor, i;
    factors = [];
    threeDigitFactors = [];
    possibleFactor = 1;
    sqrt = Math.sqrt(number);
    while (possibleFactor <= sqrt) {
        if (number % possibleFactor === 0) {
            factors.push(possibleFactor);

            otherPossibleFactor = number / possibleFactor;
            if (otherPossibleFactor !== possibleFactor && String(otherPossibleFactor).length === 3) {
                factors.push(otherPossibleFactor);
            }
        }
        possibleFactor += 1;
    }
    
    for (i = 0; i < factors.length; i += 1) {
        if (String(factors[i]).length === 3) {
            threeDigitFactors.push(factors[i]);
        }
    }

    return threeDigitFactors;
    // return threeDigitFactors.sort((a, b) => a - b); // return sorted array
}
// test
// getThreeDigitsFactors(10000);
// --> [100, 125, 200, 250, 400, 500, 625]

// check if any two three digit factors multiply to give the original number

function checkThreeDigitMultiple(number) {
    'use strict';
    var i, j, checkSet, returnSet;
    checkSet = getThreeDigitsFactors(number);
    returnSet = [];
    for (i = 0; i < checkSet.length; i += 1) {
        for (j = 0; j < checkSet.length; j += 1) {
            if (checkSet[i] * checkSet[j] === number) {
                // return number + ' = ' + checkSet[i] + ' * ' + checkSet[i]; // if you want only one answer
                returnSet.push(checkSet[i] + " * " + checkSet[j]);
            }
        }
    }
    return returnSet;
}
// test
// checkThreeDigitMultiple(20000);
// --> ["100 * 200", "200 * 100", "125 * 160", "160 * 125"] // not perfect (repeats), but works.

// main function
function checkPalindromeMultiple() {
    'use strict';
    // create a set of palindrome numbers between 100000 and 998001; check only 6-dig palindromes
    var str, palindromeSet, i, j;
    palindromeSet = [];
    for (i = 998001; i > 100000; i -= 1) { // start with the largest
        str = String(i);
        if (str.charAt(0) === str.charAt(str.length - 1) && str.charAt(1) === str.charAt(str.length - 2)) {
            if (str.length % 2 !== 0 || str.charAt(2) === str.charAt(3)) {
                palindromeSet.push(i);
            }
        }
    }
    
    // check the threeDigitFactors of the palindromeSet
    for (j = 0; j < palindromeSet.length; j += 1) {
        if (checkThreeDigitMultiple(palindromeSet[j]).length !== 0) {
            return palindromeSet[j];
        }
    }
}
// test
checkPalindromeMultiple();
// --> 906609


// This guy's code put my essay to shame but I'm still proud of the multi-use functions, and my code is actually faster. https://gist.github.com/msnider/4315679

function palindromic() {
// console.time('timeend'); // test runtime
    'use strict';
    var max, i, k, n, s;
    max = 0;
    for (i = 999; i > 99; i -= 1) {
        for (k = 999; k > 99; k -= 1) {
            n = i * k;
            s = String(n); // this, or
            // s = '' + n; // this converts a number to a string
            if (s === s.split("").reverse().join("") && n > max) {
                max = n;
            }
        }
    }
    return max;
}
// test
palindromic();
// console.timeEnd('timeend');

File: py/e4.py
# Akachukwu Obi, edited Oct 23, 2023
# Project Euler #4 - Largest Palindrome Number


# Solution 1: factorization

import math

def getThreeDigitMultiples(number = 580085):
    """returns 3-digit multiples of a number"""
    factors = []
    possibleFactor = 1
    limit = math.sqrt(number)
    
    while possibleFactor <= limit:
        if number % possibleFactor == 0: # if it is a factor
            factors.append(possibleFactor)
            otherPossibleFactor = number // possibleFactor
            if otherPossibleFactor != possibleFactor:
                factors.append(otherPossibleFactor)
        possibleFactor += 1
        
    threeDigitMultiples = []
    for i in factors:
        if len(str(i)) == 3 & len(str(number // i)) == 3:
            threeDigitMultiples.append(i)
    
    return sorted(threeDigitMultiples) # sorts the set 
# print(getThreeDigitMultiples()) # [583, 995]

def isPalindrome(number):
    """returns true if number reads the same forwards and backwards"""
    num_str = str(number) # converts number to string
    return num_str == num_str[::-1]

def checkPalindromeMultiple():
    """returns largest palindrome number with two 3-digit multiples"""
    for i in range(998001, 10000, -1):
        if isPalindrome(i) and len(getThreeDigitMultiples(i)) != 0:
            return i
        
print(checkPalindromeMultiple()) #906609
# took 0.015s runtime


# Solution 2: using python's string reverse check 

def isPalindrome(number):
    """returns true if number reads the same forwards and backwards"""
    num_str = str(number) # converts number to string
    return num_str == num_str[::-1]

def largestPalindromeProduct():
    """find largest palindrome product of two 3-digit numbers"""
    max_product = 0
    for i in range(999, 99, -1): # count backwards
        for j in range(i, 99, -1): # starting j from i avoids duplicates
            product = i * j
            if product <= max_product:
                break # leave inner loop once we find max value
            if isPalindrome(product):
                max_product = product # update max value
    return max_product

print(largestPalindromeProduct()) # 906609
# took 0.00075 s

# Solution 2 updated for divisibility by 11
def isPalindrome(number):
    """returns true if number reads the same forwards and backwards"""
    num_str = str(number) # converts number to string
    return num_str == num_str[::-1]

def largestPalindromeProduct():
    """find largest palindrome product of two 3-digit numbers
        every even-digited palindrome is divisible by 11 """
    max_product = 0
    for i in range(999, 99, -1): # count backwards
        if i % 11 != 0: # focus initial values on multiples of 11
            continue
        for j in range(999, i-1, -1): 
            product = i * j
            if product <= max_product:
                break # leave inner loop once we find max value
            if isPalindrome(product):
                max_product = product # update max value
    return max_product

print(largestPalindromeProduct()) # 906609
# took 7.32e-05 s